Update to Features of mote-api.py

Bit-River opened the pull request:
- Optional security measure to restrict IP access.

- Added a Larson scanner effect, rainbow effect (from SenseHat api), and a Tie-Dye effect.

- 2-Mote configuration type: based on two motes meeting with the ends in the middle, USB ports at the far ends.

Bit-River committed:
Update mote-api.py

Added a Larson scanner effect, rainbow effect (from SenseHat api), and a Tie-Dye effect.

Optional security measure to restrict IP access.

Bit-River committed:
Update mote-api.py

Bit-River committed:
Update mote-api.py

Gadgetoid commented on python/examples/mote-api.py:
```diff
@@ -39,7 +52,7 @@ def hex_to_rgb(value):
 def mote_on(status):
-    for chan in range(4):
+    for chan in range(1,5,1):
         r, g, b = status['colour'][chan + 1]
```
Need to drop the `+ 1` in this loop, otherwise it would be writing channels `2, 3, 4, 5`

Bit-River replied:
Thanks, I'll change these

Gadgetoid commented on python/examples/mote-api.py:
```diff
@@ -150,6 +262,219 @@ def set_brightness(channel, br):
+## 2-Mote - Clearing Back, then All-Off
+@app.route(baseurl + version + '/off_animated_2_mote/<int:ch_a>/<int:ch_b>', methods=['GET'])
+def off_animated_2_mote(ch_a,ch_b):
+    global colour
```
Redundant `global colour`?

Gadgetoid commented on python/examples/mote-api.py:
```diff
@@ -150,6 +262,219 @@ def set_brightness(channel, br):
+## 4-Mote - Clearing Back, then Filling
+@app.route(baseurl + version + '/on_animated_4_mote/colour/<string:c>', methods=['GET'])
+def on_animated_4_mote(c):
+    global colour
```
Redundant `global colour`?

Gadgetoid commented on python/examples/mote-api.py:
```diff
@@ -150,6 +262,219 @@ def set_brightness(channel, br):
+
+
+## 2-Mote - Double Light Traveliing Towards then Away - x 2
+@app.route(baseurl + version + '/cylon_duo_2_mote/colour/<string:c>/<int:ch_a>/<int:ch_b>', methods=['GET'])
```
Could there be a more extensible way of storing the mappings between URL and animation effects? Right now there's a lot of boilerplate repetition.

Gadgetoid commented on python/examples/mote-api.py:
```diff
@@ -51,13 +64,134 @@ def mote_on(status):
+    for i in range(15,-1,-1):
+        mote.set_pixel(channel,i,r,g,b)
+        mote.show()
+        time.sleep(pause_time_default)
```
`time.sleep` is blocking, and would cause the web request to hang until the animation effect has completed. Perhaps these effects should be threaded by default- that means they could also run indefinitely- and a new request could stop/change the effect.

I'd suggest a worker thread for this; constantly running in the background to manage the effects and looking at either an effect queue- populated by requests- or at a single `current_effect` value.

Since they also cloud the original purpose of the `mote-api.py` I'd probably also include the effects in a separate Python file (or multiple) and find a way to register them that handles creating the correct routes, etc.

I would likely do this by creating a `MoteEffect` class and deriving each effect from it.

Bit-River replied:
Those threading changes might be a fair bit out of my current skill range.

Right now I'm using the pauses as part of the animation effects, so some type of wait has to be there.

I'll get up to speed on making a MoteEffect class.

Gadgetoid reviewed (commented):
There are some great ideas here, thank you, and I think with a bit of work they could make some nice improvements to the Mote API that others could build upon. I've made a few suggestions alongside the code, catching a couple of bugs and making recommendations for how the effects could be better implemented.

Doing things like storing effects in separate Python files, and deriving them from an effects class makes it much easier for others to build and contribute their own effects.

Also, threading the animation of effects means the API is always responsive and returns instantly- effects can then either trump the last effect immediately, or be pushed onto a queue to play through in order.

Bit-River reviewed (commented):
Thanks for the changes.

Forgot to undo those range steps!

I'll fix up, and look into MoteEffect classes.

Cheers for the Motes!

Without them I wouldn't have started picking up Python, bet that still shows ;)

Bit-River committed:
Thinned code, made general animation effect defintion

To do:

* Make into a class

Bit-River committed:
Update mote-api.py

Gadgetoid commented:
You've got pretty far for someone "picking up" Python :D To have made working changes and proposed a PR with good ideas is a good start. Keep it up!

Bit-River committed:
New api animation call, 'larsonswipe'

Larson Swipe Call, Mote Sticks Changed Sequentially.

Altered Meaning to ch_selection.

Stick Order is used, e.g. '4300' - means the 4th Stick, and then the 3rd, and  '1342' - 1st, 3rd, 4th, then the 2nd.

Call example: /larsonswipe/1300/1/00ff00/0.1/0

Bit-River committed:
Updated functions to take repeat-loop values

Bit-River committed:
Update mote-api.py

Bit-River committed:
Added rgb version of calls 

Accomodates different input sources from Home Assistant, e.g. rgb sliders

Bit-River commented:
I'm happy enough with where this is feature-wise now.

Next step, I'll change it into a python class.

I've updated my post on Home Assistant about controlling the Motes, tidied things up a bit there:

https://community.home-assistant.io/t/mote-light-leds-versatile-variable-and-discrete-plus-home-assitant-switches-groups-api/11712

Bit-River committed:
Update mote-api.py

Cut functions in half, added simple state answer response, moved Rainbow and Tiedye out of api call section

Bit-River committed:
Update mote-api.py

Functions Complete. Mote Effect functions called from new file 'MoteEffects.py'.

Bit-River commented:
Here's the Mote Effects file to go with the above updated API:

https://github.com/Bit-River/Hass-API-testing/blob/master/MoteEffects.py 

I'm happy enough with where this is now. I'm still not up to speed on Python Classes, so didn't want to hold up things anymore and went with modules.

As I think of new effects, I'll add them to that file.

Thanks for your tips on how to proceed, and if you've time I'd love to hear what you think of this one.

There's a little bit for me to do on the Home Assistant side, e.g. integrating it with AppDaemon - so I'll probably be looking at that for a while.

Yarr :)

Bit-River reviewed (commented):
Changes made, and functions complete.

Bit-River committed:
Update mote-api.py

Took 'led_light_status()' out, and into MoteEffects.py.

Bit-River committed:
Update mote-api.py

Moved 'led_lit()' into MoteEffects.py

Bit-River commented:
Had to move the status check into the MoteEffects.py in order to get accurate readings.

And have discovered a snag, the status check in the main API are no longer accurate.

I think it's because I have called 'mote.configure_channel' in both, and so the MoteEffects.py configure is getting precedence.

I'm not sure how to fix this.

Gadgetoid commented:
I need to find some time to try this out and read through the code :D It looks like you're heading in the right direction, though, nice work!

PaulWebster commented:
I tried the code from this post (to allow Apple Homekit control of Mote lights)
https://learn.pimoroni.com/tutorial/sandyj/using-mote-with-homekit-and-siri
The code there is not the same as in github
e.g. /on replaced by one that requires channel number or "all".
So this breaks the homebridge integration.
The original API calls could be retained - and then (for example) a /on should then call the new function with "all".
Alternative is to update the homebridge config file - but then that makes it harder for anyone who has followed the example and then updates the mote scripts.

Also - the returns from API when used by HomeKit/Homebridge will, presumably have to provide the best guess reply on the assumption that all pixels are in the same state (which they would be if only set via Homekit).
Perhaps provide each channel as a different device to allow a Mote stick to be individually controlled - but probably not required.

The MoteEffects.py needs to be added to the pull request.

Gadgetoid commented:
Maintaining backwards compatibility is a tricky business, and this code has been on my to-review list for a long, long time!

Perhaps we could move forward by preserving the old `mote-api.py` and renaming this to `mote-api-v2.py` with the caveats mentioned by @PaulWebster documented, or the appropriate backwards compatibility added in?

Gadgetoid commented:
Possibly relevant, but @Godley had the idea to bake an API right into the base library with Flask Blueprints for convenience. 

See discussion here: https://github.com/pimoroni/scroll-phat-hd/issues/26

See the WIP here: https://github.com/pimoroni/scroll-phat-hd/pull/41

It would be great if the Mote API could follow this same pattern. Perhaps this gives us the separation from the old conflicted `mote_api.py` we need to move forward with this new and shiny stuff?

Gadgetoid commented:
@Bit-River I've just noticed that `MoteEffects.py` is not included in your PR. If you're still keen to work on this, then alongside renaming your API to `mote-api-v2.py`, could you include that file? Thank you!

At this point I think this constitutes a project and should be moved into `/projects/mote_api_v2/` in much the same way that Unicorn Paint is a "project" in the Unicorn HAT HD repo: https://github.com/pimoroni/unicorn-hat-hd/tree/master/projects/unicornpaint

This will keep Mote API and its dependencies tucked away in a tidy location and separates them a little from the basic examples.

Harking back to my comment in Nov, I'd still like to build API functionality into the library which a "Mote Effects" project could then use to great... effect.

Gadgetoid commented:
Jan 2018 :astonished: 

I honestly don't know where the years go. I'm closing this for now since it's cluttering up my todo list, is over four years old, and I think Mote has long since gone EOL anyway so I can't spend all that much time on it.

PaulWebster commented:
Still being sold - with 10 items listed.
https://shop.pimoroni.com/products/mote

If it is going EOL then I'd be interested in warehouse clearance discount sale ;)

Gadgetoid commented:
@PaulWebster discounts should appear in a couple of weeks.  For, uh, no specific reason :birthday: :eyes:

PaulWebster commented:
Not spotted a discount on this yet (plenty of tempting other stuff though)

https://shop.pimoroni.com/products/mote

and still flagged as "new"
